python/dgl/mock_sparse2/reduction.py

```diff
@@ -0,0 +1,322 @@
+"""DGL sparse matrix reduce operators."""
+from typing import Optional
+
+import torch
+
+from .sparse_matrix import SparseMatrix
+
+
+def reduce(A: SparseMatrix, dim: Optional[int] = None, rtype: str = "sum"):
+    """Compute the reduction of non-zero values in sparse matrix A along
+    the given dimension :attr:`dim`.
+
+    If :attr:`dim` is None, it reduces all the elements in the sparse
+    matrix. Otherwise, it reduces on the row (``dim=0``) or column (``dim=1``)
+    dimension, sproducing a tensor of shape ``(A.shape[1], ) + A.val.shape[1:]``
+    or ``(A.shape[0],) + A.val.shape[1:]``.
+
+    The reduction does not count zero values. If the row or column to be
+    reduced does not have any non-zero value, the result will be 0.
+
+    Parameters
+    ----------
+    A : SparseMatrix
+        Sparse matrix
+    dim : int, optional
+        The dimension to reduce.  Must be either 0 (by rows) or 1 (by columns).
+
+        If not given, the reduction will happen across all non-zero entries.
+    rtype: str
+        Reduction type, one of ``['sum', 'smin', 'smax', 'smean', 'sprod']``,
+        representing taking the sum, minimum, maximum, mean, and sproduct of the
+        non-zero entries.
+
+    Returns
+    ----------
+    Tensor
+        Reduced tensor
+
+    Examples
+    ----------
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([1, 1, 2])
+    >>> A = create_from_coo(row, col, val, shape=(4, 3))
+    >>> print(A.reduce(rtype='sum'))
+    tensor(4)
+    >>> print(A.reduce(0, 'sum'))
+    tensor([2, 0, 2])
+    >>> print(A.reduce(1, 'sum'))
+    tensor([1, 3, 0, 0])
+    >>> print(A.reduce(0, 'smax'))
+    tensor([1, 0, 2])
+    >>> print(A.reduce(1, 'smin'))
+    tensor([1, 1, 0, 0])
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([[1, 2], [2, 1], [2, 2]])
+    >>> A = create_from_coo(row, col, val, shape=(4, 3))
+    >>> print(A.reduce(reduce='sum'))
+    tensor([5, 5])
+    >>> print(A.reduce(0, 'sum'))
+    tensor([[3, 3], [0, 0], [2, 2]])
+    >>> print(A.reduce(1, 'smin'))
+    tensor([[1, 2], [2, 1], [0, 0], [0, 0]])
+    >>> print(A.reduce(0, 'smean'))
+    tensor([[1, 1], [0, 0], [2, 2]])
+    """
+    return torch.ops.dgl_sparse.reduce(A.c_sparse_matrix, rtype, dim)
+
+
+# pylint: disable=W0622
+def sum(A: SparseMatrix, dim: Optional[int] = None):
+    """Compute the sum of non-zero values in sparse matrix A along
+    the given dimension :attr:`dim`.
+
+    If :attr:`dim` is None, it reduces all the elements in the sparse matrix.
+    Otherwise, it reduces on the row (``dim=0``) or column (``dim=1``) dimension,
+    sproducing a tensor of shape ``(A.shape[1], ) + A.val.shape[1:]`` or
```

Comment:
Empty space here after `(A.shape[1],`, but not `(A.shape[0],)`, better be consistent